test/test_main.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+
+import unittest
+import logging
+import os
+import sys
+
+from unittest.mock import (
+    patch,
+)
+
+# This relies on the expected repository structure to find
+# a path to the charm under test.
+JUJU_CHARM_DIR = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)),
+    'charms/main'
+)
+
+del sys.path[0]
```

Comment:
I tried to avoid a side-effect here but so long as we are using `python3 -m unittest` we cannot use `if __name__ == '__main__'` here to avoid unconditional modification of sys.path. The reason is that, for example, `import juju.main` will try to do `import charm` which depends on having `<JUJU_CHARM_DIR>/lib` in `sys.path` with `<JUJU_CHARM_DIR>/lib/charm.py` available for import.

An alternative to this is using `subprocess` for every test case but it requires some mechanism to communicate charm execution output to the parent process that runs the test.

I am open to suggestions on this one.